src/Testing/src/AssemblyTestLog.cs

```diff
@@ -216,20 +217,37 @@ public static AssemblyTestLog Create(Assembly assembly, string baseDirectory)
         return new AssemblyTestLog(loggerFactory, logger, baseDirectory, assembly, serviceProvider);
     }
 
-    public static AssemblyTestLog ForAssembly(Assembly assembly)
+    public static AssemblyTestLog ForAssembly(Assembly assembly) => ForAssembly(assembly, cleanupBaseDirectory: false);
+
+    public static AssemblyTestLog ForAssembly(Assembly assembly, bool cleanupBaseDirectory)
     {
         lock (_lock)
         {
-            if (!_logs.TryGetValue(assembly, out var log))
+            if (_logs.TryGetValue(assembly, out var log))
+            {
+                if (cleanupBaseDirectory && log._cleanupState == CleanupState.None)
+                {
+                    log._cleanupState = CleanupState.CleanupRequested;
```

Comment:
Thinking about multiple test classes executing their tests in parallel, should this be bullet-proof e.g. using `Interlocked.CompareExchange(...)` (and a full-sized `int` `enum`)❔

Reply:
You could always lock in `ReportTestFailure`. Although per-instance lock might be better than the static `_lock`. I think the best options is to avoid locking or interlocking entirely by turning `_cleanupState` into two bools: `bool _wasTestFailure` and `bool _cleanupRequested`.

Based on my reading, neither of these bools could ever transition from `true` to `false`, so there's no reason for any comparisons before setting either of these to `true`. `Dispose` can then delete the directory only if `_cleanupRequested && !_wasTestFailure`.

Reply:
That'll work @halter73. I merged them into an `enum` primarily to make updates atomic but that didn't work out very well 😉